Roomnames

smurfix opened the pull request:
<!-- Keep the title short & concise so your mom can understand it -->
#### Brief overview of PR changes/additions
This patch adds the ability to display optional room names below the rooms on the map.

#### Motivation for adding to Mudlet
I want to see some more info on the map than a color (which I need for other uses) or a letter/symbol.

#### Other info (issues closed, discussion etc)

This patch adds a couple of "magic" parameters to the map's user data:
* `room.ui_showName` (global map): enables showing the checkbox "Names" below the map
* `room.ui_showName` (individual room): show the room's name below its square
* `room.ui_nameFont` (global map): display font to use for the names
* `room.ui_nameOffset` (individual room): x- and y-offsets to move the room name, to avoid collisions with its neighbours. If only one coordinate is given treat as Y offset because that's most commonly needed. Offsets are in font height units.

TODO: `room.ui_nameSize` (font size relative to the room's boxes, default 1), per-room font and size

smurfix committed:
Map: Add an option to show room names below their IDs.

This is a fairly straightforward extension.

TODO: use a separate font.

SlySven commented on src/ui/mapper.ui:
```diff
@@ -6,8 +6,8 @@
     <y>0</y>
-    <width>403</width>
-    <height>387</height>
+    <width>613</width>
```
This is quite a bit wider - and depending on the size constraints could be problematic for some users - see if you can get by without changing these values...

smurfix replied:
Meh, I didn't change this, that was the qt creator all by itself. in fact I did compensate by shortening the texts. will take another look.

SlySven replied:
Depending on what Git software you use you could have prevented it going into the commit and then reset that unwanted change back out of the codebase...

SlySven commented on src/T2DMap.cpp:
```diff
@@ -900,6 +909,15 @@ inline void T2DMap::drawRoom(QPainter& painter, QFont& roomVNumFont, QPen& pen,
+        painter.save();
+        QColor roomIdColor;
+        painter.setPen(QPen(QColor(Qt::white)));
+        painter.setFont(roomVNumFont);
```
I'd check whether this is a good font instance to use - IIRC I adjust it's size dynamically so that the largest room number in the currently shown area will just fit within the room square (so will be different if all the rooms in the area are less than, say, 10,000 compared to maps with room numbers in the millions) - so the corresponding room name text will be dependent on that as well.

You'd be better advised to work out the maximum length (remember the room names are not confined to ASCII characters so counting the `QChar`s in the text is NOT sufficient) to determine how much space the text will take up on screen - and adjust the font size appropriately! Also the font used for the room numbers may not be adequate to show all room names...

smurfix replied:
I'l add a separate option for room names and I'll figure out more reasonable values for the font sizes.

SlySven commented on src/T2DMap.cpp:
```diff
@@ -900,6 +909,15 @@ inline void T2DMap::drawRoom(QPainter& painter, QFont& roomVNumFont, QPen& pen,
+    if (mShowRoomName && areRoomIdsLegible) {
+        painter.save();
+        QColor roomIdColor;
+        painter.setPen(QPen(QColor(Qt::white)));
```
What about if the map background is not a dark colour - you will want to consider the map background colour (The `(QColor) mpHost->mBgColor_2` variable accessible in the `(void) T2DMap::paintEvent(QPaintEvent*)` method and use `Qt::white` if the `QColor::lightness()` is less than 128 or `Qt::black` if it above.

smurfix replied:
will fix

SlySven commented on src/T2DMap.cpp:
```diff
@@ -900,6 +909,15 @@ inline void T2DMap::drawRoom(QPainter& painter, QFont& roomVNumFont, QPen& pen,
+    // Do we need to draw the room name:
+    if (mShowRoomName && areRoomIdsLegible) {
+        painter.save();
+        QColor roomIdColor;
```
This isn't used at present and since the colour needed will be the same throughout the whole paint event you'd be marginally faster to set the value for this at the top of the method (as soon as it have been decided to draw a map rather than the error message about not having a map or valid user room number) as I note in the next/previous (GitHub remembers the order of comment entry rather than the comment position so *next* may not be next in practice! :roll_eyes: ) - and then use that colour variable to set the pen colour.

SlySven reviewed (commented):
There is potential here - but what is to happen when an area is set for ***Grid-mode***?

smurfix committed:
basic support for a separate font for room names

smurfix committed:
Use black room names on light background

smurfix committed:
Room name display and font sizing fixes

Room name display is now independent of room number display.
The cut-off for the font size is lower (4 instead of 7).

Room names are scaled as if the number 88 would be printed in the room
box, which works for all practical purposes.

Font size discovery is now done in floating point, and by scaling with
factors 1.2 (up) and 1.05 (down) which is faster *and* looks way better.

smurfix committed:
Don't show room names in grid mode

smurfix committed:
Tweak font size calculation

Ensure that we start with a reasonable font size

smurfix committed:
Added functions to get/set a room name's offset

float x/y, relative to the room rectangle (and in its units).

TODO: document in the wiki:

getRoomNameOffset(id)
setRoomNameOffset(id, x_offset, y_offset)

smurfix committed:
Save file version updated to 21

Added:
* global: room name font + size adjustment
* per room: label offset

SlySven commented on src/T2DMap.cpp:
```diff
@@ -900,6 +909,24 @@ inline void T2DMap::drawRoom(QPainter& painter, QFont& roomVNumFont, QPen& pen,
+        painter.save();
+        QColor roomNameColor;
+        if (mpHost->mBgColor_2.lightness() > 127) {
+            roomNameColor = QColor(Qt::black);
```
This might be a case where the ternary operator is a tad more efficient - i.e. replace the whole `if` with:
```CPP
QColor roomNameColor((mpHost->mBgColor_2.lightness() > 127) ? Qt::black : Qt::white);
```

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -1645,6 +1645,22 @@ int TLuaInterpreter::centerview(lua_State* L)
     }
 }
 
+// Documentation: https://wiki.mudlet.org/w/Manual:Lua_Functions#updateMapView
```
Why do you need this - isn't the existing **[updateMap()](https://wiki.mudlet.org/w/Manual:Lua_Functions#updateMap)** function sufficient?

smurfix replied:
No, I was merely blind and didn't see that.

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -1645,6 +1645,22 @@ int TLuaInterpreter::centerview(lua_State* L)
+
+    if (!host.mpMap || !host.mpMap->mpRoomDB || !host.mpMap->mpMapper) {
+        lua_pushnil(L);
+        lua_pushstring(L, "centerview: you haven't opened a map yet");
```
On ***run time*** errors *(ones where the command is validly constructed but the current situation is not compatible with the function)* our current (though not necessarily updated in all previous code - :eyes: - oh :hankey:! There are around 40 uses like this an only one of them has had the function name taken out!!! :roll_eyes:) UI style is not to report the function that has the problem {c.f. ones where the user has provided an incorrect type of argument where we abort by using `lua_error(lua_State*)`}.

Okay  - the suggested text is `"no map present or loaded!"`  but you can, I suppose, prepend `"updateMapView: "` to it - if you do need this (see other comment).

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -11745,7 +11761,7 @@ int TLuaInterpreter::setRoomChar(lua_State* L)
         lua_pushnil(L);
         lua_pushfstring(L, "room with id %d does not exist", id);
-        return 2;
+        return 1;
```
:facepalm: :blush:

smurfix replied:
Right. That was a brain fart on my part.

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -11777,13 +11793,81 @@ int TLuaInterpreter::getRoomChar(lua_State* L)
         lua_pushnil(L);
         lua_pushfstring(L, "room with id %d does not exist", id);
-        return 2;
+        return 1;
```
:man_facepalming: :woman_facepalming: :roll_eyes:

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -11777,13 +11793,81 @@ int TLuaInterpreter::getRoomChar(lua_State* L)
+        lua_pushfstring(L, "setRoomNameOffset: bad argument #1 type (room id as number expected, got %s!)",
+                       luaL_typename(L, 1));
+        return lua_error(L);
+    } else {
```
:rotating_light: :robot: *CLazy **anti-pattern** alert!*

:information_source: The code can be made cleaner by loosing the `else` - as the code contained in *that* part can never be reached if the other part is run - which has the bonus feature of not having to separate the declaration of a variable and assigning a value to it...

I.e. you can change:
```CPP
    int id;
```
and
```CPP
    if (!lua_isnumber(L, 1)) {
        lua_pushfstring(L, "setRoomNameOffset: bad argument #1 type (room id as number expected, got %s!)",
                       luaL_typename(L, 1));
        return lua_error(L);
    } else {
        id = lua_tointeger(L, 1);
    }
```
to the simpler:
```CPP
    if (!lua_isnumber(L, 1)) {
        lua_pushfstring(L, "setRoomNameOffset: bad argument #1 type (room id as number expected, got %s!)",
                       luaL_typename(L, 1));
        return lua_error(L);
    }

    int id = lua_tointeger(L, 1);
```

SlySven replied:
And the same thing in other cases further down...

smurfix replied:
I do know that that would be cleaner. However, for all of this I simply copied the existing code. Coding style diffs are all well and good but shouldn't get mixed with bug fixes or feature changes.

I can certainly go through that code and clean up all these return-else antipatterns, but I'll do that when everything else has settled down, otherwise there'll be unnnecessary merge and/or rebase conflicts.

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -11777,13 +11793,81 @@ int TLuaInterpreter::getRoomChar(lua_State* L)
+        pR->nameOffset.setX(x);
+        pR->nameOffset.setY(y);
+        lua_pushboolean(L, true);
+        return 1;
```
In *this* case the `return` has the same value in both branches - so doesn't need to be inside the `if (...) {...} else {...}`!

SlySven commented on src/TRoom.cpp:
```diff
@@ -799,6 +800,12 @@ void TRoom::restore(QDataStream& ifs, int roomID, int version)
+                qreal x, y;
+                ifs >> x;
+                ifs >> y;
+                nameOffset = QPoint(x, y);
```
Don't you need `nameOffset = QPointF(x, y);` here - `x` and `y` are floating point numbers and `QPoint` uses (faster) integers only.

SlySven commented on src/dlgProfilePreferences.cpp:
```diff
@@ -189,6 +189,8 @@ dlgProfilePreferences::dlgProfilePreferences(QWidget* pF, Host* pHost)
     fontComboBox_mapSymbols->setToolTip(tr("<p>Select the only or the primary font used (depending on <i>Only use symbols "
                                            "(glyphs) from chosen font</i> setting) to produce the 2D mapper room symbols.</p>"));
+    fontComboBox_mapNames->setToolTip(tr("<p>Select the font used to show the 2D mapper room names.</p>"));
+    spinBox_mapNamesSizeAdj->setToolTip(tr("<p>Adjust the font size of the room names.</p>"));
```
I thought the font size was being adjusted when it was drawn beneath the rooms - so why must the size be set here? :confused:

SlySven commented on src/ui/mapper.ui:
```diff
@@ -696,7 +696,23 @@
+                <verstretch>0</verstretch>
+               </sizepolicy>
+              </property>
+              <property name="layoutDirection">
```
Not sure this property is needed - and it might be problematic if anyone tries to add support for Arabic (or is it Hebrew) locales which use `RightToLeft` - I'd manually cut it out of the file (you might be able to undo it in the Qt Designer plugin/utility by clicking the small left pointing arrow on the extreme right of the option which should reset any option back to the default - and cause it to loose the boldness in its listing in that tool:

Look at the layoutDefault setting for this control (which won't be in `.ui` file):
![image](https://user-images.githubusercontent.com/6163092/88406868-2caedf80-cdc9-11ea-9912-d053c22bbddc.png)

See how it has been bolded after I change it but put it back to the default (for my current locale as it happens):
![image](https://user-images.githubusercontent.com/6163092/88407151-97f8b180-cdc9-11ea-85d2-af987b44521c.png)

However clicking on that (now red) arrow reverts/resets it (and disables the arrow and turns that grey):
![image](https://user-images.githubusercontent.com/6163092/88407369-eefe8680-cdc9-11ea-9a49-a753ec4cf2fa.png)

smurfix replied:
Yeah, mis-click in the designer. I'll go through the designer files' changes and fix these parts.

SlySven commented on src/ui/profile_preferences.ui:
```diff
@@ -42,7 +42,7 @@
      </property>
      <property name="currentIndex">
-      <number>0</number>
+      <number>5</number>
```
:information_source: Try to remember to select the first tab before saving the dialogue/form - though there is now code in place to force the first tab to be shown (except in a particular special case) this change happens a lot because people forget it and it gets logged into the git history when ideally it shouldn't be...

SlySven commented on src/ui/profile_preferences.ui:
```diff
@@ -1984,77 +1984,6 @@
          </layout>
         </widget>
        </item>
-       <item row="1" column="0">
```
I am going to climb onto my soapbox (and Vadi will probably try to push me off! :laughing: ) but if possible it is really helpful if you are editing things in a `QGridLayout` if you open it in the `Plain Text editor` afterwards and put the `<item row="'x'" column="'y'">` items back into row then column order. The Qt Designer plugin/utility tends to put them back in a random order after editing which makes the git diff much harder (almost impossible) to review for that file.

 Fortunately the plain text editor has the capability to fold (conceal lower - more indented - blocks of XML text) so you can click on the down/right triangles in the margin (which don't show up well in my current colour scheme) to change that and then select each item and cut and paste it to rearrange them into order - {though it gets a bit funny with maintaining the right order when pasting an item as the FIRST one and auto expands any that are pasted as the last one - so you have to be a bit tricksy in the sequence of operations. {Hint: if you are going do this - save a backup of the `.ui` file first to cover your back...!}

![image](https://user-images.githubusercontent.com/6163092/88408841-02125600-cdcc-11ea-922c-0ef4fdbcfa20.png)

SlySven commented on src/ui/profile_preferences.ui:
```diff
@@ -3380,6 +3409,15 @@
                     </color>
                    </brush>
                   </colorrole>
+                  <colorrole role="PlaceholderText">
```
Aaargh! - Another bug with the Qt Designer plugin / utility - is that it assumes that the `.ui` files that it makes are only going to be used by the current Qt version (specifically the `QUILoader` class) of the libraries and it will happily put in details that are un-parsable and will break on previous versions of Qt - and the `PlaceholderText` element of a `QPalette` is one of them - it was introduced in, I think Qt 5.12 but our code must be compilable going back to Qt **5.11** versions - therefore you will need to manually (with the plain text editor) going to have to excise all the `<colorrole role="PlaceholderText">`...`</colorrole>` elements in the file... {though the previous comment's hint I gave about the element folding will help you to do this. :smiling_imp: }

SlySven replied:
That is why the `.7` job of the Travis Linux CI build is failing - it specifically runs on the oldest Qt version we support to test for this sort of thing. :grin:

smurfix replied:
Fixing. Yes I do know how to use an XML editor. ;-)

SlySven commented on src/ui/profile_preferences.ui:
```diff
@@ -5306,9 +5443,6 @@
   <tabstop>comboBox</tabstop>
   <tabstop>mMapperUseAntiAlias</tabstop>
   <tabstop>checkBox_showDefaultArea</tabstop>
-  <tabstop>fontComboBox_mapSymbols</tabstop>
```
Please put these back in the right place - though of course changes to this form will of course clash with PR #3929 .

SlySven reviewed (commented):
Not sure about the sizing of the font being both specified as a user option and than also done for each name text - that seems counter productive. As for font selection you might keep things simpler by using the same font that is used for the info box at the top of the screen. The only reason I made it possible to select a (and force the use of a single) font for the room symbols is that that is a particular case where a map maker might have their own custom font for the map (they may even use Private Use glyphs) - I suspect this level of control is not necessary for showing the room names and it might make things simpler for you and the end-user to not go to the same extent for them...

smurfix committed:
remove LayoutDirection directives

smurfix committed:
Re-sorted the grid layout items in ui/profile_prefs

smurfix commented:
OK, I think I got all review comments handled.

smurfix commented:
As my dialog changes conflict with PR #3929, I'll do the work to reconcile the two, whenever that PR (or this one) gets merged.

vadi2 reviewed (commented):
This is a great addition, but it needs some more work:

![Selection_696](https://user-images.githubusercontent.com/110988/89117737-74f09080-d4a0-11ea-9143-b98ae478a854.png)

smurfix commented:
You need to spread your map to fit the room names, and you need to stagger the labels.

Spreading a map is already a Mudlet feature. My patch adds new functions to lookup and change the label's offset, unimaginatively named `get/setRoomNameOffset`.

Actually modifying your existing map to look good with room names, and/or auto-staggering the names of existing rooms / new rooms as you create them is outside the scope of this patch. Everybody's map is different, mine looked this way (meaning, with default room distance 5) way before I decided to write the roomnames patch. One possible extension would be a "how far does this name stretch" inquiry function, but I decided to hold off on that until people actually indicate a need for this – my mapping doesn't require it, simple string length heuristic is more than sufficient even with a variable-width font.

The default vertical distance between the room square and its name is fixed, so it you increase the "Rooms" control below the map it'll look a whole lot better.
![Screenshot from 2020-08-02 15-35-24](https://user-images.githubusercontent.com/236571/89124245-ca936000-d4d5-11ea-8aad-c05b060e405c.png)

I plan to add some "how to make beautiful maps with room names" text to the wiki as soon as this patch goes in.

vadi2 commented:
It means you turn on some functionality in Mudlet, and it's "broken by default". That's not the way to go.

Maps already have a room label functionality - if you'd like to auto-layout this yourself, have you looked into those? Improving them to state where they can emulate this could be a way to go - fixes your usecase while not overpromising and underdelivering for others.

smurfix commented:
I don't think it's "broken by default". Maps laid out without regard for room names typically have no space for room names. That's a fact. I cannot get around that.

You want space for labels, you need to rearrange your rooms so that the labels fit, either manually or with some additional tooling.
I don't think it's fair to block this extension until somebody gets around to writing these tools, esp. since there's not much of an incentive to do so if the basic features required for the tools aren't in Mudlet.

That being said, the default position for a name should be directly below that room's square no matter how large that square is, thus they should shift down when you enlarge the squares. I'll fix that.

vadi2 commented:
Google maps doesn't have to move rivers and mountains in order to fit map labels, so I don't think we have to do it here either. There's always space on the edges. If we're to have this in Mudlet, turning on the display should not be a disappointment of "oh god, this looks horrible" (and I know people who'll say this!).

You'd like something custom for yourself and that's OK - we already have labels, why not use them?

smurfix commented:
Real-world mappers have rather sophisticated layout algorithms for their names which are somewhat overkill in a MUD which mainly has a rectangular grid. We're talking multiple weeks of development effort to code iterative layouting that looks as well as what a human can do intuitively. Nobody is going to do that IMHO.

Or, my approach, the room names get a default location and somebody (i.e. someone who knows their way around Mudlet's mapper better than I do) adds some code to shift them around a bit, in order to teach east/west adjacent room labels to not overlap. If some name is still in the way, well, click on its room, press Control+up/down/left/right to move the label around until placed reasonably well, done. Macros to do so are dead simple, at least as soon as somebody figures out how to capture Control-up and Control-down, right now that doesn't work. :-/

Yes we already have labels. They're meant to be small and to in a room's box. A MUD's "short" room names are almost never small enough to do so comfortably, and if you shorten them you're back to manual intervention.

What do you want me to do -- teach the existing labels to reflect the MUD's room name if there's no explicit room symbol? Maybe wrap that text so that it fits into the room's box with a larger font? (Would require some auto-hyphenation, which will probably do strange things with a MUD's fantasy names *shrug*) All possible, but IMHO somewhat independent of the ability to position the room name outside that box, wherever you want/need it.

smurfix committed:
Merge remote-tracking branch 'origin/development' into roomnames

smurfix committed:
Fix room name placement

directly under the room's rectangle, no strange offset

smurfix commented:
OK, vertical placement of room names is now fixed: they're directly below the boxes.

I cannot do anything about colliding names. That's an order of magnitude more difficult than auto-placing rooms so they won't collide, which Mudlet doesn't do either. You can move the labels around easily, by adding macros for Control-Keypad-2648 or the cursor keys (except that Mudlet can't trap up/down for some reason). I might even be persuaded to add a patch that does this.

Yes of course we already have manual labels. But nobody is going to manually type and place labels for ten thousand rooms. Manual labels have a different use case (on my maps they're mainly used for "here's the way to region Foo" and "this is the town of Quux", readable even when zoomed far out) and there's no reason we can't have both.

smurfix commented:
OK … is there anything else I should do to get this accepted?

NB, another reason for this patch is that names are attached to their rooms. If you create "normal" labels for every room then you can no longer move or shrink or … a group of rooms without making a mess of things because all the labels stay where they are.

vadi2 commented:
I think that's a problem we should solve with labels then. It's not like these labels are attached to the room visually anyway? This patch functionality could be replicated by a Lua script to create labels at x,y,z of every room...

The bigger problem is that the out of the box experience  will be poor for players that want to see room names.

Additional problems are - questions like why do we have two types of labels which are similar? Both are movable, neither are attached to the map in any way...

There's more problems than help right now here.

Happy to hear others opinions on this, if you use the mapper give this a go and add your opinion.

smurfix commented:
Huh? "my" room labels are of course attached to the map, they move with the rooms.

Kebap commented:
I for one welcome Smurfix's suggestion to display room names directly adjacent to each room painted.

Then again, I also agree with Vadi's sentiment to not have an ugly look and feel for each new player, unless they find Mudlet wiki and follow some (albeit simple) steps to correct this. 

Mudlet out-of-the-box must be nice to newbies, not require additional effort just to be okay, actually that's the whole purpose of Mudlet 5.0 roadmap.

Therefore, can I suggest we use a different field in the room data to gather the displayed texts from? For most players, this field shall be empty, therefore no text is shown. Only if they explicitly put the room name (or any other desired text) in there, then Mudlet will paint the text and keep it at the room visually, etc. This can be automated with mapper scripts where desired.

vadi2 commented:
Something intermediate that doesn't promise to work out of the box to the user - and in Mudlet, everything should - would be good :+1: . The other field idea is a good one, we also need a solution for the UI. You press a button -> you expect it to work.

smurfix commented:
Adding a different field (and renaming the method to position it) would be no problem at all,. Whatever it takes to get something like this merged.

As for the UI, my initial solution would be a "show/hide room name" menu item in the map that simply sets the displayed name to the mapname (or clears it if it's set). Should be disabled if room names are not displayed of course. We can do more elaborate things in Lua.

vadi2 commented:
I think we can automatically show the UI toggle _if_ the custom roomnames field are found to be populated in the map.

How we store them userdata field of every room? I see SlySven has been using a `system.blah` and `audit.blah` convention - so something like `room.ui_showName`? The tricky thing here is to make it obvious why is it different from just the `setRoomName()` function.

edit: actually on second thought we don't need to duplicate the data, so just a boolean true/false in userdata would be enough and then the actual room name is used.

OK, we're near a workable solution here. I want to try this again to see how does editing the labels works, since that's essential to enable this.

smurfix committed:
Merge remote-tracking branch 'origin/development' into roomnames

smurfix commented:
`room.ui_showName` (instead of another bool in `TRoom` and another two Lua functions) works for me.
I could also store the label position (relative to the bottom center of the room rect) in `room.ui_nameOffset` or similar (ditto), and a per-room font size adjustment would probably be a good idea too.

That way I could drop the map file version uptick.

smurfix committed:
Add userdata flag "room.ui_showName" indicating whether to show room name

smurfix committed:
Move room label position to userdata

smurfix committed:
Cleanup

The flag whether to display room labels (both globally and per-room)
is now stored in userdata.

Reverted:

ad6d3 Re-sorted the grid layout items in ui/profile_prefs
76834 remove LayoutDirection directives
5d81e Save file version updated to 21

TODO: flipping the global flag crashes for no good reason.

smurfix committed:
Crash workaround

Not inlining "setUserDataBool" triggers a double-free bug
(Debian amd64, Qt 5.14.2-3, gcc 10.2.0-9).

I'm investigating this; in the meantime, this patch should be an
acceptable workaround.

vadi2 commented on src/TRoomDB.cpp:
```diff
@@ -1326,3 +1330,36 @@ void TRoomDB::setAreaRooms(const int areaId, const QSet<int>& roomIds)
+bool setUserDataBool(QMap<QString, QString>& userData, const QString& key, bool value)
+{
+    if (value) {
+        userData[key] = QString("1");
```
```suggestion
        userData[key] = QStringLiteral("1");
```

vadi2 commented on src/TRoomDB.cpp:
```diff
@@ -1326,3 +1330,36 @@ void TRoomDB::setAreaRooms(const int areaId, const QSet<int>& roomIds)
+    if (value) {
+        userData[key] = QString("1");
+    } else {
+        userData[key] = QString("0");
```
```suggestion
        userData[key] = QStringLiteral("0");
```

vadi2 commented on src/mudlet-lua/lua/GUIUtils.lua:
```diff
@@ -2066,6 +2064,49 @@ function resizeMapWidget(width, height)
+-- @param flag (bool)
+function setRoomNameVisible(room, flag)
+  assert(type(room) == 'number', 'setRoomNameVisible: bad argument #1 type (room ID as number expected, got '..type(room)..'!)')
+  assert(type(flag) == 'bool', 'setRoomNameVisible: bad argument #2 type (flag as bool expected, got '..type(flag)..'!)')
```
```suggestion
  assert(type(flag) == 'boolean', 'setRoomNameVisible: bad argument #2 type (flag as boolean expected, got '..type(flag)..'!)')
```

vadi2 committed:
Update src/mudlet-lua/lua/GUIUtils.lua

smurfix committed:
Appease codefactor

vadi2 commented:
Can the room offset be moved via the UI?

smurfix commented:
You mean, as in dragging+dropping the label? No, I didn't think about how to do that yet.

In my own code I shift them around with control+keypad+2468. I'd love to do it with the arrow keys instead but up/down isn't currently forwarded to the key bindings.

vadi2 commented:
OK, that is fine, we can use the aliases for this.

Just one last bit, can you hide the `Names` UI option unless the map has some names enabled? 

For the non-technical users, an option that doesn't seem to do anything on the surface will be less confusing - and the technical ones will figure out how to get it to show.

smurfix commented:
Meh, that would require walking through all the rooms in the current area (and/or the whole map) to figure out whether any of those are set.

Alternate idea: show the checkbox only when the global userData entry exists. A mapper which shows roomnames can then set that if it doesn't exist, which will enable the checkbox.

Both of these strategies require a way to trigger a dialog re-draw. :-/

vadi2 commented:
I think that will be fine.

We don't have to redraw the box, just mark the element as hidden or visible: https://doc.qt.io/qt-5/qwidget.html#visible-prop

smurfix commented:
OK, the problem then boils down to extending `updateMap` since that's going to be called anyway when you enable this feature. On the other hand I don't want to add special-case code for this to `setMapUserData`.

Will do.

smurfix committed:
Merge remote-tracking branch 'origin/development' into roomnames

smurfix committed:
Hide showRoomNames checkbox if no room name infrastructure exists

as exhibited by map userData "room.ui_showName" not being present at all

smurfix commented:
Implemented; if the `room.ui_showName` key is not present in the map's userData then the checkbox is hidden.

I dunno if we should add Lua functions to control this or just document what the `room.*` userData items do; if the former, I'll do that in another PR.

Kebap commented:
For now, let's just document all this, including alias to move labels, etc. 
Then see how much traction it actually gets, before we add more features.

smurfix committed:
Merge remote-tracking branch 'origin/development' into roomnames

vadi2 commented:
Could you update the PR description with the end solution?

I'll see about updating the mapping scripts we have to handle this addition.

smurfix commented:
done

vadi2 reviewed (approved):
Thanks a lot for this improvement!

smurfix commented:
Less effort than writing a mapper that works the way I want it to from scratch 😎

vadi2 commented:
:) could you document this on the wiki? Bonus point points if you could also update the generic mapping script to handle editing this.

smurfix commented:
Wiki: done.

I'm not using the generic mapper, so letting me modify it is probably of limited usefulness. 🤔
